docs/features/file-based-programs-vscode.md

```diff
@@ -0,0 +1,97 @@
+
+## High-level IDE layer changes
+
+The following outline is for "single-file" scenarios. We are interested in expanding to multi-file as well, and this document will be expanded in the future with additional explanation of how to handle that scenario.
+
+### Prerequisite work
+- MSBuildHost is updated to allow passing project content along with a file path. (https://github.com/dotnet/roslyn/pull/78303)
+- LanguageServerProjectSystem has a base type extracted so that a FileBasedProgramProjectSystem can be added. The latter handles loading file-based program projects, and makes the appropriate call to obtain the file-based program project XML. (https://github.com/dotnet/roslyn/pull/78329)
+
+### Expected flow
+- User opens a new instance of VS Code.
+- User opens a .cs file.
+- C# extension starts up.
+- `LspWorkspaceManager.GetLspDocumentInfoAsync` is called for the document.
+- Code path `_lspMiscellaneousFilesWorkspace?.AddMiscellaneousDocument` is hit.
+- Synchronously, a "miscellaneous project" is synthesized via `MiscellaneousFileUtilities.CreateMiscellaneousProjectInfoForDocument`, and the document within it is returned from `GetDocumentInfoAsync`.
+    - There is also a "metadata as source" code path. This *should* just be a fairly simple easy-out path we can take in the new workspace kind, as it is in the existing misc files workspace.
+- Asynchronously, a file-based program project is created and passed off to the MSBuild host.
+    - When this project finishes loading, the "miscellaneous project" is replaced with the "file-based program project".
+    - The document is also checked for whether it it is a file-based program per [Heuristic](#heuristic).
+    - If it is a file-based program, then the bit `hasAllInformation: true` is passed along in the resulting `ProjectAttributes`.
+    - If it is not a file-based program, then `hasAllInformation: false` is passed for the same field. This allows syntax diagnostics and Quick Info to work for known symbols, which improves productivity without adding "noise" for files where we do not know all the context.
+    - A notification is fired off which causes `GetLspDocumentInfoAsync` to run again and get a new `Document` from a now fully up-to-date workspace.
+- Asynchronously, the current solution or list of projects is also loaded in the main "host workspace". We may discover that a file we were creating a *file-based program project* for, is actually included in an ordinary project. In this case, when `GetLspDocumentInfoAsync` runs again, and we find the file in the host workspace, we take care to call `_lspMiscellaneousFilesWorkspace.TryRemoveMiscellaneousDocument` with the file. This unloads any *miscellaneous project* or *file-based program project* associated with the file.
+
+### Heuristic
+The IDE considers a file to be a file-based program, if:
+- It has a `#!` directive such as `#!/usr/bin/env dotnet run`, or,
```

Comment:
Is this a good heuristic? `#!` can appear in scripts too.

Reply:
> It might be also allowed in all C# files in the future, there is no reason to disallow it.

What does it meant to a "regular" C# in this case? Something that's still a single file program or something else?

Reply:
I meant a normal C# file that's part of a project-based program. In other words, we don't have to disallow `#!` in any files, I think (we don't interpret it anyway, shells do, based on whatever user puts there, so it should be user's choice where to put it). But currently we disallow `#!` in project-based programs (and allow it in scripts and file-based programs).

Reply:
I have removed `#!` as part of the heuristic. Thanks